Allow use of overloaded methods when using LoggerMessage

JakeYallop opened the pull request:
Fixes #61814

Computes a unique name for `LoggerMethod`s, allowing for overloaded methods to be used with the source generator.

JakeYallop committed:
Compute unique method names for generated code

Calculates and assigns a unique LoggerMethod name as required, and uses
this new unique name for generating types in the source generator. This
allows using the LoggerMessageAttribute on methods that share the same
name but have different method signatures.

Fix #61814

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=64573) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=64573) before we can accept your contribution.<br/><br/>:x: JakeYallop [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=64573)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=64573) it.</sub>

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+
+        [LoggerMessage(EventId = 1, Level = LogLevel.Trace, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, string p0);
+
```
nit: multiple blank lines

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -510,6 +510,23 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
+                            {
+                                if (methods.ContainsKey(lm.Name))
+                                {
+                                    var currentCount = methods[lm.Name];
```
nit: as per coding style, please use int not var

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Microsoft.Extensions.Logging.Generators.Tests.TestClasses
+{
+    internal static partial class OverloadTestExtensions
```
nit: might want to rename filename to `OverloadTestExtensions` to keep the filename and class name in sync

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+        public static partial void M0(ILogger logger, string p0);
+
+
+        [LoggerMessage(EventId = 2, EventName = "M1Custom", Level = LogLevel.Trace, Message = "M1{p0}")]
```
Aren't the first two `M0` enough? are you testing something new in the methods `M1` below? If not, then can perhaps remove.

JakeYallop replied:
My thinking for this was to verify that custom event names are honoured for each individual overload of a logging method. As the implementation is now, it doesn't do a lot, but if the implementation was to change in future as to how EventNames are processed in the parser, then this test would highlight that issue.

maryamariyan reviewed (approved):
LGTM. Nice work.

maryamariyan commented:
This could be merged once you get a chance to apply the minor feedback in this PR.

JakeYallop committed:
Rename MethodOverloadTestExtensions to OverloadTestExtensions

Also resolves other PR feedback, an extra blank line and incorrect use
of an implicit type.

JakeYallop commented:
@maryamariyan

I've updated the PR based on feedback. The only thing left is the last two LoggerMessage overload tests - I've written a comment with my thinking on that - let me know if you still want me to remove them,

maryamariyan commented:
@JakeYallop Thanks for the quick turnaround.

I would still be tempted to remove the last two overloads. event name uniqueness is not currently being respected and I was thinking we should fix that first and just leave this PR to focus on overloads for now.

JakeYallop committed:
Remove un-needed overload tests

maryamariyan commented:
Failures in CI seem unrelated and I also found them in https://github.com/dotnet/runtime/pull/64593 and https://github.com/dotnet/runtime/pull/64641

maryamariyan commented:
/backport to release/6.0

eerhardt commented:
@maryamariyan - are we sure this needs to be backported to 6.0? It doesn't feel like this limitation is critical to fix in 6, but I guess I'd like your opinion/thoughts.

maryamariyan commented:
I assume it seems like a blocker to developers wanting to migrate to the logging source generator approach. Looking back at the issue doesn't seem like there'd been a lot of hits with this bug as much as there were for the `@` prefixed issue. So I'd definitely service that one.
